client/state/partner-portal/products/selectors.ts

```diff
@@ -0,0 +1,46 @@
+import { PartnerPortalStore, APIProductFamilyProduct } from 'calypso/state/partner-portal/types';
+// Required for modular state.
+import 'calypso/state/partner-portal/init';
+
+export function getSelectedProductSlugs( state: PartnerPortalStore ) {
+	return state.partnerPortal.products.selectedProductSlugs;
+}
+
+export function getDisabledProductSlugs(
```

Comment:
I would definitely like to memoize this as mentioned by @atanas-dev in p9oQ9f-1qT-p2#comment-2111. Even if this particular selector isn't very computationally heavy we will need a way to memoize selectors which combine data from redux with data from react-query. The memoizing will need to respond to changes both in the redux data and the react-query data.

I have been doing some testing with [create-selector](https://github.com/Automattic/wp-calypso/tree/trunk/packages/state-utils/src/create-selector). This does allow memoization based on selector parameters via the use of the `getDependants` function. However based on my testing it seems that `react-query` is returning a new reference with each call to `useQuery` even when values are cached. `createSelector` is doing a shallow comparison of the dependents, so memoizing isn't working out of the box.

I find that if I change the dependent comparison to use `isEqual` from lodash instead of `isShallowEqual` [here in createSelector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/index.ts#L97) then the selector is correctly memoized, or if I use `JSON.stringify(allProducts)` instead of `allProducts` alone then the selector correctly memoizes, for example:

```javascript
export const getTotalSelectedCost = createSelector(
	( state: PartnerPortalStore, allProducts: APIProductFamilyProduct[] ) => {
		const selectedProductSlugs = getSelectedProductSlugs( state );

		const totalSelectedProductCost = allProducts
			.filter( ( product ) => selectedProductSlugs.includes( product.slug ) )
			.map( ( product ) => product.amount )
			.reduce( ( totalCost, productCost ) => totalCost + productCost, 0 );

		return totalSelectedProductCost;
	},
	( state: PartnerPortalStore, allProducts: APIProductFamilyProduct ) => {
		// Doing this causes the selector to correctly memoize
		return [ getSelectedProductSlugs( state ), JSON.stringify( allProducts ) ];
		// Doing this causes the selector to re-run with every render
		// return [ getSelectedProductSlugs( state ), allProducts ];
	}
);
```

If it was all in the redux state tree then the shallow comparison would find the references to the products to be identical and memoization would be the clear winner. However `react-query` must be returning new references which prevents a quick shallow comparison. My thought for this particular selector is that the cost of deep comparing the arrays or stringifying them is probably a wash compared to just running the selector again.

Reply:
Thanks for further researching ways to memoize the selector. 

> it seems that react-query is returning a new reference with each call to useQuery even when values are cached. 
I find that odd - have you been able to find a clue that this really is the case?

Regardless, it would be worth doing some benchmarking to see what is faster and if memoization is actually worth it here. In any case, if we find out that memoization with `JSON.stringify` is faster than no-memoization, I'd be fine to go with that 👍